### PR TITLE
[Workspace] Filter left nav menu items according to the current workspace

### PR DESCRIPTION
**Pull request by ruanyl**
### Description
In this PR, left nav bar will be filtered by the feature configuration of current workspace.

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6153

## Screenshot

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/486382/03c7a1c3-6a5e-4557-bd13-ef3d927ecea9

Accessing app which is NOT configured by the workspace will display "Application Not Found" page
image

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/public/chrome/nav_links/nav_links_service.ts [hunk 1]
```diff
@@ -53,6 +53,13 @@ export interface ChromeNavLinks {
    */
   getNavLinks$(): Observable<Array<Readonly<ChromeNavLink>>>;
 
+  /**
+   * Get an observable for the current link updaters. Link updater is used to modify the
+   * nav links, for example, filter the nav links or update a specific nav link's properties.
+   * {@link LinksUpdater}
+   */
+  getLinkUpdaters$(): BehaviorSubject<LinksUpdater[]>;
```
Comment (Hailong-am) [inline src/core/public/chrome/nav_links/nav_links_service.ts:61]: should we update the doc https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/chrome/README.md 
and https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/public.api.md as well?

Comment (ruanyl) [reply src/core/public/chrome/nav_links/nav_links_service.ts:61]: good point, I will update [README.md](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/chrome/README.md), `public.api.md` is automatically generated.

**Review by SuZhou-Joe — APPROVED** [review]

File: src/plugins/workspace/public/utils.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+export const featureMatchesConfig = (featureConfigs: string[]) => ({
+  id,
+  category,
+}: {
+  id: string;
+  category?: AppCategory;
+}) => {
+  let matched = false;
+
+  for (const featureConfig of featureConfigs) {
```
[the remark this answers is not in the document]
Comment (ruanyl) [reply src/plugins/workspace/public/utils.ts:33]: Sure, updated.

**Comment by seraphjiang** [comment]
@xluo-aws  @ruanyl @BionIT @Flyingliuhub  @bandinib-amzn   @kgcreative @xeniatup @zengyan-amazon 

Looks like we are going to have to global contexts: Workspace at left nav, DataSource at top navi if both are enabled

have we discussed or made the decision between Workspace and DataSource?

I assume DataSource always belongs to Workspace. if we all agree about it,  from information hierarchy's perspective, should workspace selector being put on top nav bar. Once cx select workspace, then they could select datasource?

I feel we need to build both experience into future playground to get board feedback asap. 

what do you think?

image

**Comment by ruanyl** [comment]
@seraphjiang 
> have we discussed or made the decision between Workspace and DataSource?
> I feel we need to build both experience into future playground to get board feedback asap.

Agree on that, and yes, the current behavior of setting workspace and data source are more like setting two global contexts. However, base on my understanding so far, they may have correlations. For example, data source available in one workspace may not be available in another workspace.

I think we need to take this into consideration from UX design's perspective as early as possible. @kgcreative @lauralexis @xeniatup

**Comment by SuZhou-Joe** [comment]
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/application/application_service.tsx#L231 
Happen to find a filter by using capability, not sure if we can leverage capability service to filter the navLinks.

**Comment by ruanyl** [comment]
> https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/application/application_service.tsx#L231 Happen to find a filter by using capability, not sure if we can leverage capability service to filter the navLinks.

Good point! Let me try it out

**Review by SuZhou-Joe — DISMISSED** [review]
LGTM

**Comment by ruanyl** [comment]
@SuZhou-Joe I changed the implementation to use `application.registerAppUpdater` to update the app to `inaccessible`, with this change, if accessing the apps which are not configured into a workspace, the "Application Not Found" page will be displayed.

File: src/plugins/workspace/public/utils.ts [hunk 3]
```diff
@@ -55,3 +60,42 @@ export const featureMatchesConfig = (featureConfigs: string[]) => ({
...
+   * When workspace has no features configured, all apps are considered to be accessible
+   */
+  if (!workspace.features) {
+    return true;
+  }
+
+  /**
+   * The app is configured into a workspace, it is accessible after entering the workspace
+   */
+  const featureMatch = featureMatchesConfig(workspace.features);
```
Comment (Flyingliuhub) [inline src/plugins/workspace/public/utils.ts:78]: `featureMatch` or `featureMatcher`, I feel that "matcher" more accurately refers to the mechanism responsible for performing the evaluation. Just my two cents. :)

Comment (ruanyl) [reply src/plugins/workspace/public/utils.ts]: @Flyingliuhub Updated :)

**Comment by SuZhou-Joe** [comment]
@ruanyl sounds great, thanks for the change!

**Review by SuZhou-Joe — DISMISSED** [review]

**Comment by SuZhou-Joe** [comment]
Some lines are missing unit test coverage, could you please write some to cover that so that all checks can pass for this PR. :)

**Review by SuZhou-Joe — APPROVED** [review]

**Review by Flyingliuhub — APPROVED** [review]

**Comment by Hailong-am** [comment]
I have concern about mark those links as `inaccessible` for now, before we move index pattern/ saved objects page out, user have no ways to import data and create index pattern. That means user can't start create visualization and dashboards in a fresh workspace

**Comment by ruanyl** [comment]
> I have concern about mark those links as `inaccessible` for now, before we move index pattern/ saved objects page out, user have no ways to import data and create index pattern. That means user can't start create visualization and dashboards in a fresh workspace

You'r right, maybe we should wait for index pattern to be moved out before merging this PR. For saved objects page, you mean saved object management page? Do we have upcoming PR/task about that?

**Comment by ruanyl** [comment]
@xluo-aws In this PR, it will forbid user to access the applications which are not configured in a workspace, even if user access directly via url. I believe this is the desired behavior, but this is different from what we implemented earlier which we just hide the nav links from ui in left nav menu. Do you have any concern with this?

**Comment by SuZhou-Joe** [comment]
Perhaps we can hide it from nav in this PR, and adjust the logic to make these inaccessible in the PR that moves index pattern out of data management.

**Comment by xluo-aws** [comment]
> @xluo-aws In this PR, it will forbid user to access the applications which are not configured in a workspace, even if user access directly via url. I believe this is the desired behavior, but this is different from what we implemented earlier which we just hide the nav links from ui in left nav menu. Do you have any concern with this?

No concern,  it's the right behavior.

**Comment by ruanyl** [comment]
> I have concern about mark those links as `inaccessible` for now, before we move index pattern/ saved objects page out, user have no ways to import data and create index pattern. That means user can't start create visualization and dashboards in a fresh workspace

Since Index pattern will remain in "Dashboards Management" and  "Dashboards Management" is available in workspace, so this is no longer a concern, is that correct?

**Review by Hailong-am — APPROVED** [review]

**Pull request merged**